firewood/src/proof.rs

```diff
@@ -146,25 +146,31 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
     fn locate_subproof<'a>(
         &self,
         mut key_nibbles: NibblesIterator<'a, 0>,
-        encoded_node: &[u8],
+        node: NodeType,
     ) -> Result<(Option<SubProof>, NibblesIterator<'a, 0>), ProofError> {
-        let items: Vec<Encoded<Vec<u8>>> = bincode::DefaultOptions::new()
-            .deserialize(encoded_node)
-            .map_err(ProofError::DecodeError)?;
-
-        match items.len() {
-            EXT_NODE_SIZE => {
-                let mut items = items.into_iter();
-                let decoded_key: Vec<u8> = items.next().unwrap().decode()?;
+        match node {
+            NodeType::Leaf(n) => {
+                let cur_key = &n.path().0;
+                // Check if the key of current node match with the given key
+                // and consume the current-key portion of the nibbles-iterator
+                let does_not_match = key_nibbles.size_hint().0 < cur_key.len()
+                    || !cur_key.iter().all(|val| key_nibbles.next() == Some(*val));
 
-                let decoded_key_nibbles = Nibbles::<0>::new(&decoded_key);
+                if does_not_match {
+                    return Ok((None, Nibbles::<0>::new(&[]).into_iter()));
+                }
 
-                let (cur_key_path, term) =
-                    PartialPath::from_nibbles(decoded_key_nibbles.into_iter());
-                let cur_key = cur_key_path.into_inner();
+                let data = n.data();
```

Comment:
Nit:
```suggestion
                let encoded = n.data().to_vec();
```

Then below becomes:
```rust
                let sub_proof = SubProof {
                    encoded,
                    hash: None,
                };
```